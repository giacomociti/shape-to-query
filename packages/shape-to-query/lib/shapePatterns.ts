import { BaseQuad } from 'rdf-js'
import $rdf from '@zazuko/env'
import { sparql, SparqlTemplateResult } from '@tpluscode/sparql-builder'
import { UNION } from '@tpluscode/sparql-builder/expressions'

export interface ShapePatterns {
  whereClause: string | SparqlTemplateResult
  constructClause: BaseQuad[]
  childPatterns?: ShapePatterns[]
  /**
   * Patterns to inserted into a UNION block
   */
  unionPatterns?: string | SparqlTemplateResult
}

export const emptyPatterns: ShapePatterns = {
  whereClause: '',
  constructClause: [],
  childPatterns: [],
}

export function flatten(...patterns: ShapePatterns[]): ShapePatterns {
  if (patterns.every(sp => sp === emptyPatterns)) {
    return emptyPatterns
  }

  const whereClause = patterns.reduce((prev, next) => sparql`${prev}\n${next.whereClause}`, sparql``)
  const unionPatterns = patterns.reduce((prev, next) => !next.unionPatterns ? prev : sparql`${prev}\n${next.unionPatterns}`, sparql``)

  return {
    whereClause,
    unionPatterns,
    constructClause: unique(patterns.flatMap(p => p.constructClause)),
    childPatterns: patterns.flatMap(p => p.childPatterns || []),
  }
}

function unique(...construct: BaseQuad[][]): BaseQuad[] {
  const set = $rdf.termSet(construct.flatMap(arr => arr))
  return [...set]
}

export function union(...patterns: ShapePatterns[]): ShapePatterns {
  const nonEmpty = patterns.filter(p => p !== emptyPatterns)

  if (nonEmpty.length === 0) {
    return emptyPatterns
  }

  const unionedPatterns = nonEmpty.filter(({ whereClause }) => whereClause)

  if (unionedPatterns.length === 1) {
    return unionedPatterns[0]
  }

  const unionedBgps = unionedPatterns.map(({ unionPatterns, whereClause }, _, arr) =>
    arr.length > 1 && unionPatterns ? sparql`${unionPatterns}\n${whereClause}` : whereClause,
  )

  return {
    constructClause: unique(nonEmpty.flatMap(p => p.constructClause)),
    whereClause: sparql`${UNION(...unionedBgps)}`,
    childPatterns: nonEmpty.flatMap(p => p.childPatterns || []),
  }
}
