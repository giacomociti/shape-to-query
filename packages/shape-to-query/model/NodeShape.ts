import { Variable } from 'rdf-js'
import { isResource } from 'is-graph-pointer'
import { sparql, SparqlTemplateResult } from '@tpluscode/sparql-builder'
import { FocusNode } from '../lib/FocusNode'
import { emptyPatterns, flatten, ShapePatterns, union } from '../lib/shapePatterns'
import { VariableSequence } from '../lib/variableSequence'
import { PropertyShape } from './PropertyShape'
import { Target, TargetNode } from './target'
import { ConstraintComponent } from './constraint/ConstraintComponent'
import { OrConstraintComponent } from './constraint/or'

interface Parameters {
  focusNode: FocusNode
  variable: VariableSequence
}

export interface NodeShape {
  buildPatterns(arg: Parameters): ShapePatterns
  buildConstraints(arg: Parameters & { valueNode: Variable }): string | SparqlTemplateResult
}

export default class implements NodeShape {
  constructor(
    public readonly targets: ReadonlyArray<Target>,
    public readonly properties: ReadonlyArray<PropertyShape>,
    public readonly constraints: ReadonlyArray<ConstraintComponent>,
  ) {
  }

  buildPatterns(arg: Parameters): ShapePatterns {
    let { focusNode, variable } = arg
    let targets: ShapePatterns = emptyPatterns
    if (arg.focusNode.termType === 'Variable') {
      targets = union(...this.targets.flatMap(target => target.buildPatterns(<any>{ focusNode, variable })))
    }
    if (this.targets.length === 1 && this.targets[0] instanceof TargetNode) {
      const [target] = this.targets
      if (isResource(target.nodes)) {
        targets = emptyPatterns
        focusNode = target.nodes.term
      }
    }

    const properties = union(...this.properties.map(p => p.buildPatterns({ focusNode, variable })))
    const logical = union(
      ...this.constraints
        .filter((l): l is OrConstraintComponent => l instanceof OrConstraintComponent)
        .flatMap(l => l.inner.map(i => i.buildPatterns({ focusNode, variable }))),
    )

    return flatten(targets, properties, logical)
  }

  buildConstraints(arg: Parameters & { valueNode: Variable }): string | SparqlTemplateResult {
    const constraints = this.constraints.map(c => c.buildPatterns(arg)).filter(Boolean)
    if (constraints.length === 0) {
      return ''
    }

    return sparql`${constraints}`
  }
}
