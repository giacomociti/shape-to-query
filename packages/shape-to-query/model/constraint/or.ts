import { SparqlTemplateResult, sparql } from '@tpluscode/sparql-builder'
import { UNION } from '@tpluscode/sparql-builder/expressions'
import { sh } from '@tpluscode/rdf-ns-builders'
import { NodeShape } from '../NodeShape.js'
import { ModelFactory } from '../ModelFactory.js'
import ConstraintComponent, { assertList, Parameters, PropertyShape } from './ConstraintComponent.js'

export class OrConstraintComponent extends ConstraintComponent {
  constructor(public readonly inner: ReadonlyArray<NodeShape>) {
    super(sh.OrConstraintComponent)
  }

  static * fromShape(shape: PropertyShape, factory: ModelFactory) {
    const ors = shape.get(sh.or) || []

    for (const or of ors) {
      assertList(or)
      yield new OrConstraintComponent(or.list.map(p => factory.nodeShape(p)))
    }
  }

  buildPropertyShapePatterns(arg: Parameters): SparqlTemplateResult {
    const propExpr = arg.propertyPath ? sparql`${arg.focusNode} ${arg.propertyPath} ${arg.valueNode} .` : ''

    const inner = this.inner
      .map(i => i.buildConstraints(arg))
      .filter(Boolean)
      .map(innerResult => sparql`
        ${propExpr}
        ${innerResult}
      `)
    return sparql`${UNION(...inner)}`
  }
}
