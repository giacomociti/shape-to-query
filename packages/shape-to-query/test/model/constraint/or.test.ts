import { expect } from 'chai'
import { OrConstraintComponent } from '../../../model/constraint/or.js'
import { NodeShape } from '../../../model/NodeShape.js'
import { emptyPatterns } from '../../../lib/shapePatterns.js'

describe('model/constraint/or', () => {
  before(() => import('../../sparql'))

  it('combines all inner constraints where in UNION', () => {
    // given
    const foo: NodeShape = {
      buildConstraints: () => 'foo constraint',
      buildPatterns: () => emptyPatterns,
    }
    const bar = {
      buildConstraints: () => 'bar constraint',
      buildPatterns: () => emptyPatterns,
    }
    const baz = {
      buildConstraints: () => 'baz constraint',
      buildPatterns: () => emptyPatterns,
    }
    const constraint = new OrConstraintComponent([foo, bar, baz])

    // when
    const whereClause = constraint.buildPatterns(<any>{})

    // then
    expect(whereClause).to.equalPatterns(`{
      foo constraint
    } UNION {
      bar constraint
    } UNION {
      baz constraint
    }`)
  })

  it('skips constraints which returned empty', () => {
    // given
    const foo: NodeShape = {
      buildConstraints: () => 'foo constraint',
      buildPatterns: () => emptyPatterns,
    }
    const bar = {
      buildConstraints: () => '',
      buildPatterns: () => emptyPatterns,
    }
    const baz = {
      buildConstraints: () => 'baz constraint',
      buildPatterns: () => emptyPatterns,
    }
    const constraint = new OrConstraintComponent([foo, bar, baz])

    // when
    const whereClause = constraint.buildPatterns(<any>{})

    // then
    expect(whereClause).to.equalPatterns(`{
      foo constraint
    } UNION {
      baz constraint
    }`)
  })
})
