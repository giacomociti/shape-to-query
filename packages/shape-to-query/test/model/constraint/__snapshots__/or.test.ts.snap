// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model/constraint/or combines all inner constraints where in UNION 1`] = `
Array [
  Object {
    "patterns": Array [
      Object {
        "patterns": Array [
          Object {
            "triples": Array [
              Object {
                "object": Variable {
                  "value": "o",
                },
                "predicate": Variable {
                  "value": "p",
                },
                "subject": Variable {
                  "value": "s",
                },
              },
            ],
            "type": "bgp",
          },
        ],
        "type": "group",
      },
      Object {
        "patterns": Array [
          Object {
            "triples": Array [
              Object {
                "object": Variable {
                  "value": "c",
                },
                "predicate": Variable {
                  "value": "b",
                },
                "subject": Variable {
                  "value": "a",
                },
              },
            ],
            "type": "bgp",
          },
        ],
        "type": "group",
      },
      Object {
        "patterns": Array [
          Object {
            "triples": Array [
              Object {
                "object": Variable {
                  "value": "z",
                },
                "predicate": Variable {
                  "value": "y",
                },
                "subject": Variable {
                  "value": "x",
                },
              },
            ],
            "type": "bgp",
          },
        ],
        "type": "group",
      },
    ],
    "type": "union",
  },
]
`;

exports[`model/constraint/or skips constraints which returned empty 1`] = `
Array [
  Object {
    "patterns": Array [
      Object {
        "patterns": Array [
          Object {
            "triples": Array [
              Object {
                "object": Variable {
                  "value": "o",
                },
                "predicate": Variable {
                  "value": "p",
                },
                "subject": Variable {
                  "value": "s",
                },
              },
            ],
            "type": "bgp",
          },
        ],
        "type": "group",
      },
      Object {
        "patterns": Array [
          Object {
            "triples": Array [
              Object {
                "object": Variable {
                  "value": "z",
                },
                "predicate": Variable {
                  "value": "y",
                },
                "subject": Variable {
                  "value": "x",
                },
              },
            ],
            "type": "bgp",
          },
        ],
        "type": "group",
      },
    ],
    "type": "union",
  },
]
`;
